### DISCUSSION TITLE
TriggerDagRunOperator does not trigger dag on subsequent run even with reset_dag_run=True

### DISCUSSION THREAD

**Discussion by don1uppa** [apache/airflow#24548] [General]
### Apache Airflow version

2.2.5

### What happened

I have a dag that starts another dag with a conf.   
I am attempting to start the initiating dag a second time with different configuration parameters.   I it to start the other dag with the new parameters.

## What you think should happen instead

It should use the new conf when starting the called dag

### How to reproduce

See code in squsequent message

### Operating System

windows and ubuntu

### Versions of Apache Airflow Providers

N/A

### Deployment

Docker-Compose

### Anything else

**Work around for now is to delete the previous "child" dag runs.**

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
Not obvious at all. Could you please be more specific about scenario and reproduction - including logs showing what's going on. 

Converting into discussion until more info provided.

**Reply by don1uppa** [reply]
I created a couple of dags to show the issue - report_parms just prints the config and demo_TriggerDagRunOperator_issue starts it twice with different parameters.   Notice the second execution produces the same results using the first execution parameters.

Log looks like this:
[2022-06-19, 23:57:24 UTC] {logging_mixin.py:109} INFO - --------------- see parameters below --------------
[2022-06-19, 23:57:24 UTC] {logging_mixin.py:109} INFO - **{'feed_type': 'feed_type-1', 'file_id': 'file_id-1', 'file_name': 'file_name-1'}**

Perhaps it isn't even running it a second time, I may need to modify the example.


[demo_TriggerDagRunOperator_issue.zip](https://github.com/apache/airflow/files/8936647/demo_TriggerDagRunOperator_issue.zip)
.

**Comment by don1uppa** [comment]
I worked on this some more, it seems the problem is the dag does not get run the second time unless it has a different trigger_run_id, changing the execution_date has no effect.
[triggerDagIssue.zip](https://github.com/apache/airflow/files/8940559/triggerDagIssue.zip)

see code with comments.

**Comment by potiuk** [comment]
Yes of course. You need to do backfill to re-run task that already completed.

**Reply by don1uppa** [reply]
What is the purpose ot this flag then?

reset_dag_run ([bool](https://docs.python.org/3/library/functions.html#bool)) – Whether or not clear existing dag run if already exists. This is useful when backfill or rerun an existing dag run. When reset_dag_run=False and dag run exists, DagRunAlreadyExists will be raised. When reset_dag_run=True and dag run exists, existing dag run will be cleared to rerun.

**Reply by don1uppa** [reply]
Can we turn this back into a bug report please?

**Reply by don1uppa** [reply]
Please, this appears stuck here in limbo. @potiuk

**Reply by don1uppa** [reply]
Please, this appears stuck here in limbo. @potiuk

**Reply by potiuk** [reply]
I think you should investigate further. Looking several time at your issue, I am unable to figure out what I should do to reproduce your issue quickly.  Maybe I am stupid and someone else is better for that, or maybe you should spend more time in explaining your case - I do not know.

But I think you have a problem with your way of doing your stuff and instead of providing reproduction steps. you are just dumping the DAGs but you do not show the exact observations you experience. The messages above indicate some disconected observations but they do not show the exact steps:

1. I did that
2. I saw that
3. I expected something else
4. my observation is this...

For example when you run a trigger with "reset_dag_run=True" - you should see the log file:

```
              self.log.info("Clearing %s on %s", self.trigger_dag_id, parsed_execution_date)
 ```

Do you see it? Is it something you observe? what did you do to get to the point you observe.

I think the best way to move this forward is that - rather than uploading the DAGs, you should explain step-by-step what you did and what logs you saw along the way. Just dumping dags is not enough. Remember, you get the software for free. This is not a "help-desk" - people here are trying to help if you have problems, but they are doing it in their free time. it's your task to make sure that you explain your problem in the way that people who look at it are able to understand what you are doing without spending a lot of their own time on trying to figure out what you are trying to do.  You can watch my talk about empathy from the Airflow Summit -  https://www.youtube.com/watch?v=G6VjYvKr2wQ&list=PLGudixcDaxY2LxjeHpZRtzq7miykjjFOn&index=55 - in OSS it does not work to demand answer, it works when you try to make your own effort in explainigng 

Again - maybe I am too stupid to understand what you explain. in which case - please open your issue again, add all relevant details - maybe someone else will pick it up. 

Also you might not be aware that when you have free software that you pay 0 USD for, raising a "bug" or "discussion" does not matter. If you do not describe your issue in clear enough way, there is high probablity nobody will look at it - and then there is no matter if it is an issue or discussion. If no-one will take care of it, nothing will happen wiht it. The "limbo" you mentioned is more because you have not spend enough time on making sure that whoever looks at it understands your problem, than the fact how it is classified. It's you who raise your problem who are interested in fixing it. Your answers are as good as the description you get.

**Reply by don1uppa** [reply]
Thank you for responding, I'll try to be clearer.

Problem: The TriggerDagRunOperator does not trigger the target dag on subsequent run even with the reset_dag_run option set to true.

I provided a set of DAGS to reproduce the problem.

**Scenario**: The first dag called "**demo_TriggerDagRunOperator_issue**' in my provided example uses the  TriggerDagRunOperator to start the second DAG in my provided example called "**print_params**".  

The  first time the **demo_TriggerDagRunOperator_issue**  dag is executed it starts the second dag "**print_params**" as expected  from task "**starter**" .   

Then the second task  "starter2" is executed using a different configuration.  

The log shows it worked but looking at the  second dag "**print_params**" it only shows the configuration of the first execution.   I would expect the print_parm dag would have 2 attemps not just one and it would have the correct configuration printed.

I hope you can follow this.   
 Let me know please.

I did spend time figuring out the issue and creating  those DAGS to  demonistrate the issue.

If you can point me the the source code in the repo  I might be able to debug the issue and provide a fix.
[
[print_parms.log](https://github.com/apache/airflow/files/9032224/print_parms.log)
[starter_task.log](https://github.com/apache/airflow/files/9032225/starter_task.log)
[starter_task2.log](https://github.com/apache/airflow/files/9032227/starter_task2.log)

**Reply by don1uppa** [reply]
![image](https://user-images.githubusercontent.com/33661098/176982553-7bc2bcc3-3cae-4865-8c07-5c1cdae9d3e6.png)
![image](https://user-images.githubusercontent.com/33661098/176982644-324447e8-bcab-427b-a6c3-a178a4500cc7.png)

**Reply by potiuk** [reply]
Ah I see now what you wanted to achieve. This explanation was much clearer. I actually did not need to look at the atttached DAGs. I just needed a clear description of what you want to achieve - and the picture of yours with explanation, finally managed to do that. Thanks for that.

This is misunderstanding how DagRun reset works. The "reset" is not "recreate". This is where you got things wrong. Dag Run configuraiton is immutable, once it is created, re-triggering (and resetting) resets only the "execution" of it but it does not "recreate" the dag run with it's configuration. 

You simply can 't do it. If you want to run two dagruns with two different configurations, they should have different run id. There is nothing to fix. It's just how "reset" works.

If this is not clear from the documentation, I'd appreciate if you can make PR with clearer description that if you pass dag run configuration, it is only going to be used when the dag run is created, not when it is "Reset".

**Reply by justinyouwork** [reply]
hi I'm experiencing the same issue here. I also tried with different trigger_run_id for each dag trigger. It just re-run the previous run instance with the old conf. hi @don1uppa , did you find a solution?

**Comment by don1uppa** [comment]
Hi Jarek,  Thank you.   Do you think they would be open to a recreate
option as an enhancement?

On Wed, Jul 6, 2022 at 9:11 AM Jarek Potiuk ***@***.***>
wrote:

> Ah I see now what you wanted to achieve. This explanation was much
> clearer. I actually did not need to look at the atttached DAGs. I just
> needed a clear description of what you want to achieve - and the picture of
> yours with explanation, finally managed to do that. Thanls for that.
>
> This is misunderstanding how DagRun reset works. The "reset" is not
> "recreate". This is where you got things wrong. Dag Run configuraiton is
> immutable, once it is created, re-triggering (and resetting) resets only
> the "execution" of it but it does not "recreate" the dag run with it's
> configuration.
>
> You simply can 't do it. If you want to run two dagruns with two different
> configurations, they should have different run id. There is nothing to fix.
> It's just how "reset" works.
>
> If this is not clear from the documentation, I'd appreciate if you can
> make PR with clearer description that if you pass dag run configuration, it
> is only going to be used when the dag run is created, not when it is
> "Reset".
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/airflow/discussions/24548#discussioncomment-3092411>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIA2BKRCJ3O5NDQ7DN22TD3VSWAZ5ANCNFSM5ZHFB5NQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Donald Iuppa
131 Eaglesfield Way
Fairport, NY 14450
***@***.***
Ph: (585) 210-8744

JOHN 11:19-27  Jesus said to her, “I am the resurrection and the life;
whoever believes in me, even if he dies, will live, and anyone who lives
and believes in me will never die.”

**Comment by don1uppa** [comment]
Hi jerek, there is no way to delete an existing run except via the rest api
correct?

On Wed, Jul 6, 2022 at 10:08 AM Donald Iuppa ***@***.***> wrote:

> Hi Jarek,  Thank you.   Do you think they would be open to a recreate
> option as an enhancement?
>
> On Wed, Jul 6, 2022 at 9:11 AM Jarek Potiuk ***@***.***>
> wrote:
>
>> Ah I see now what you wanted to achieve. This explanation was much
>> clearer. I actually did not need to look at the atttached DAGs. I just
>> needed a clear description of what you want to achieve - and the picture of
>> yours with explanation, finally managed to do that. Thanls for that.
>>
>> This is misunderstanding how DagRun reset works. The "reset" is not
>> "recreate". This is where you got things wrong. Dag Run configuraiton is
>> immutable, once it is created, re-triggering (and resetting) resets only
>> the "execution" of it but it does not "recreate" the dag run with it's
>> configuration.
>>
>> You simply can 't do it. If you want to run two dagruns with two
>> different configurations, they should have different run id. There is
>> nothing to fix. It's just how "reset" works.
>>
>> If this is not clear from the documentation, I'd appreciate if you can
>> make PR with clearer description that if you pass dag run configuration, it
>> is only going to be used when the dag run is created, not when it is
>> "Reset".
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/apache/airflow/discussions/24548#discussioncomment-3092411>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AIA2BKRCJ3O5NDQ7DN22TD3VSWAZ5ANCNFSM5ZHFB5NQ>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>
>
> --
> Donald Iuppa
> 131 Eaglesfield Way
> Fairport, NY 14450
> ***@***.***
> Ph: (585) 210-8744
>
> JOHN 11:19-27  Jesus said to her, “I am the resurrection and the life;
> whoever believes in me, even if he dies, will live, and anyone who lives
> and believes in me will never die.”
>


-- 
Donald Iuppa
131 Eaglesfield Way
Fairport, NY 14450
***@***.***
Ph: (585) 210-8744

JOHN 11:19-27  Jesus said to her, “I am the resurrection and the life;
whoever believes in me, even if he dies, will live, and anyone who lives
and believes in me will never die.”

**Reply by potiuk** [reply]
This is the way that is recommended. There are probably other ways. For example  manually updating the DB, but we are not recommending it and if you choose this path, you have to make sure to manually update ti to match our later database schema. So there ARE other ways but they involve you taking more responsibility for it, where REST API is an API that we are committed to keep backwards compatible in Airflow 2.

And answering the other question of yours:

> Do you think they would be open to a recreate option as an enhancement?

I think it is a total abuse of what Airflow is for. Probably you cna use Excel spreadsheat to run database query if you really want, but should you do it? 

I am not sure why you want to do that - it makes very little sense in Airflow world - you keep database of DagRuns to keep history of executiion of the Dag. This is basic feature of Airflow. If you want to do something else, maybe you should not use Airflow at all. It seems it causes you more troubles than solving your problems and maybe you just need a Python script to execute your Python code? Why do you want to use Airlfow to trigger stuff if you do not want to keep history ? It's a bit puzzlling what would be your case here.

**Reply by sahilthapar** [reply]
@potiuk Added a PR to improve the docstring in TriggerDagRunOperator
#26755

**Comment by don1uppa** [comment]
Hi Jarek, That makes sense.  Shouldn't we at least get a second execution
attempt with the initial configuration though?

On Thu, Jul 7, 2022 at 4:53 PM Jarek Potiuk ***@***.***>
wrote:

> This is the way that is recommended. There are probably other ways. For
> example manually updating the DB, but we are not recommending it and if you
> choose this path, you have to make sure to manually update ti to match our
> later database schema. So there ARE other ways but they involve you taking
> more responsibility for it, where REST API is an API that we are committed
> to keep backwards compatible in Airflow 2.
>
> And answering the other question of yours:
>
> Do you think they would be open to a recreate option as an enhancement?
>
> I think it is a total abuse of what Airflow is for. Probably you cna use
> Excel spreadsheat to run database query if you really want, but should you
> do it?
>
> I am not sure why you want to do that - it makes very little sense in
> Airflow world - you keep database of DagRuns to keep history of executiion
> of the Dag. This is basic feature of Airflow. If you want to do something
> else, maybe you should not use Airflow at all. It seems it causes you more
> troubles than solving your problems and maybe you just need a Python script
> to execute your Python code? Why do you want to use Airlfow to trigger
> stuff if you do not want to keep history ? It's a bit puzzlling what would
> be your case here.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/airflow/discussions/24548#discussioncomment-3103203>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIA2BKVHO7O6UCM6F4YQDEDVS47VLANCNFSM5ZHFB5NQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Donald Iuppa
131 Eaglesfield Way
Fairport, NY 14450
***@***.***
Ph: (585) 210-8744

JOHN 11:19-27  Jesus said to her, “I am the resurrection and the life;
whoever believes in me, even if he dies, will live, and anyone who lives
and believes in me will never die.”

**Reply by potiuk** [reply]
Are you sure you did not get it ? Or you did not look hard enough?

**Reply by don1uppa** [reply]
I'm sure, here are screen shots.
![image](https://user-images.githubusercontent.com/33661098/177895742-d103f287-eddf-450d-8df4-05db56fd18c6.png)
![image](https://user-images.githubusercontent.com/33661098/177895836-dcb439e7-a979-4aec-8cd9-74e3d395e6c3.png)
![image](https://user-images.githubusercontent.com/33661098/177896000-bfe4fc3d-8518-40c9-a135-3f834b35a094.png)
![image](https://user-images.githubusercontent.com/33661098/177898692-eeef9c08-eac9-4c90-ada0-191ea786ecc3.png)

**Reply by potiuk** [reply]
But are you sure the second run actually run "TriggerDagRunOperator" ? Do you have logs showing it? What do they tell?

**Reply by potiuk** [reply]
Also did you try to experiment with reset_dag_run= False and maybe using other non-existing dags. What happens when you try ?

**Reply by don1uppa** [reply]
Here is the log of the second task "starter2
![image](https://user-images.githubusercontent.com/33661098/178320896-28c8a756-79f8-4431-b5f4-c0a993b0863c.png)

If I change the reset_dag_run=False here is the log

![image](https://user-images.githubusercontent.com/33661098/178372948-2c539703-3c98-4f33-8b31-6dcaadc3d1d2.png)

**Reply by potiuk** [reply]
the log of which run ? and what do you gather from those logs and experiments ?

**Reply by don1uppa** [reply]
# Task will sleep beyond the 10 second SLA to showcase sla_miss callback
    bash_sleep = BashOperator(
        task_id="bash_sleep",
        bash_command="sleep 180",
    )

    doTrigger1 >> bash_sleep >> doTrigger2

**I added a sleep between the first and send starter tasks so I could look at the print_params log.   I can confirm that it did not change and I only see one attempt.   With starter2 reset_dag_run= False  we get an error "Dag run exists" set to True it does not fail and it doesn't do anything.**




![image](https://user-images.githubusercontent.com/33661098/178375111-7f5cc274-ab71-4381-91ad-441f22ef7de2.png)

**Comment by potiuk** [comment]
I am confused. I see that the dag was triggered for the second time. how do you know print_params.log was not simply overwritten with the same output?

 We already agreed that the DagRun params are immutable, I do not know logic of the print_params.log but I guess it will be the same when you re-run it and parms will be kept from the first run. What makes you think it's not run but gets overwritten with exactly the same value. 

What do you expect to happen and is not happening?

**Comment by don1uppa** [comment]
I watched the log and noticed that the timestamps of the log messages did
not change.

**Comment by don1uppa** [comment]
I would  expect to a least see a second run with  new log output
timestamps.  I really think we should see a second attempt.  The first
attempt should not be removed.  I think it should behave just like you
cleared the first run.

**Reply by potiuk** [reply]
What does the starter2 log show?

**Comment by don1uppa** [comment]
No, I did not get around this issue.  I avoid trying to retrigger this
way.   I was told it is working as designed.

On Wed, Jun 28, 2023 at 2:58 PM justinyouwork ***@***.***>
wrote:

> hi I'm experiencing the same issue here. I also tried with different
> trigger_run_id for each dag trigger. It just re-run the previous run
> instance with the old conf. hi @don1uppa <https://github.com/don1uppa> ,
> did you find a solution?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/airflow/discussions/24548#discussioncomment-6308074>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIA2BKVRPVZ7G7NQMLUJAFLXNR5DPANCNFSM5ZHFB5NQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Donald Iuppa
9934 Tioga Trail
Pinckney, Michigan 48169
-- and --
131 Eaglesfield Way
Fairport, NY 14450
***@***.***
Ph: (585) 210-8744

JOHN 11:19-27  Jesus said to her, “I am the resurrection and the life;
whoever believes in me, even if he dies, will live, and anyone who lives
and believes in me will never die.”

**Comment by FFCMSouza** [comment]
I'm having the exact same issue here.
The only diference is that i don't try to change de conf of the dag_run.
When I clear the task with the trigger dag operator i see log saying that the target dag as clear, but no retry as found in the dag.
`[2023-10-07, 00:52:17 -03] {trigger_dagrun.py:173} INFO - Clearing c3po_processos_starwars on 2023-10-07 03:52:15.990912+00:00`
If the target dag is still running then the calling dag wait his execution, if is in success or failed state just return the result of the target dag.
I'm using Airflow 2.7.1.
Can anyone help me with this issue?

**Comment by brendan-morin** [comment]
I am also having the same issue on 2.2.x. From what I am seeing, I think that it is inappropriate and dismissive to suggest this is user error or behaving as designed. If for some reason the DAG will not actually be triggered by a *trigger* operator, telling the user the dag is being cleared and then silently doing nothing is not the right answer. Perhaps this is fixed in later versions and a non-issue.

I am using a custom run_id (e.g. `run_id=123`). When retriggering, the logs claim it is clearing the future dag run (e.g. "Clearing my_dag on 2023-10-13 16:57:12.268145+00:00", but I can look at the dag run_id it was supposed to clear and trigger and can confirm it does not clear it (or do anything).

It is possible that having a custom run_id may have something to do with it. I'm not on the latest, and it's been a bit since I've gone source diving (for a separate, but related issue), but I think I recall there is some logic that tries to find DAG by time range that is broken when using a non time-based run_id. In general I find that using manually specified run_ids are not consistently supported in predictable ways.

Our present workaround is to actually just delete the DAG-to-be-triggered if one with a run_id exists first, and then trigger it so Airflow thinks it's a fresh run. It's not great since we lose logs and state for the old run, but at least it's predictable and idempotent.

**Reply by potiuk** [reply]
Oh yes. I missed the last comments so let me re-convert it back to issue. Seems that there is no re-run even if no configuration change - thanks for bringing my attention to ti @BrendanJM . I am not sure either if it's fixed in any of the newer version but it would certainly help if somoene who experiences it would confirm that it is still the same in the latest version (this is where the fix will come in case someone implements it anyway, so upgrading to latest version is anyhow a good idea).

**Reply by FFCMSouza** [reply]
same issue on 2.7.1.

**Reply by crutis** [reply]
Also having this problem on 2.7.2 when using a custom operator:
Maybe the operator is calling the wrong api? I see:
```
dag.clear(start_date=parsed_execution_date, end_date=parsed_execution_date)
```
My database (for reasons beyond my control) is in local time, not UTC ... perhaps dag.clear fails because of a timezone issue?

I would expect to see something analogous to the rest api call to [clear tasks for a given dag run](https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html#operation/clear_dag_run):
```
POST https://airflow.apache.org/api/v1/dags/{dag_id}/dagRuns/{dag_run_id}/clear
```
